### PR TITLE
Standalone M3 Control Node

### PR DESCRIPTION
**Pull request by rdagher**
issue #55

### CODE REVIEW

**Comment by rienafairefr** [comment]
I took the liberty to rebase on latest master, much clearer now :-)
While I look at this more, can you verify it still works for your setup ? (I had no major merge conflicts on the code though)

File: gateway_code/control_nodes/cn_iotlabstalone/__init__.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
+# -*- coding:utf-8 -*-
```
Comment (rienafairefr) [inline gateway_code/control_nodes/cn_iotlabstalone/__init__.py]: stalone looks like a typo to me, I guess the directory should be named cn_standalonem3 ?

File: gateway_code/control_nodes/cn_iotlab/__init__.py [hunk 2]
```diff
@@ -190,7 +190,7 @@ def _wait_control_node_ready():
         So wait 2 seconds to be safe.  """
         time.sleep(2)
 
-    @staticmethod
-    def status():
+    @logger_call("Control Node : status")
```
Comment (rienafairefr) [inline gateway_code/control_nodes/cn_iotlab/__init__.py]: Yeah, this looks fine, but this is not related to standalone M3 control node IMO. Put that un another PR, with other small fixes

File: gateway_code/control_nodes/cn_iotlabstalone/__init__.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+from gateway_code.utils.ftdi_check import ftdi_check
+from gateway_code.config import static_path
+
+from gateway_code.control_nodes.cn_iotlab import ControlNodeIotlab
+
+
+LOGGER = logging.getLogger('gateway_code')
+
+
+class ControlNodeIotlabstalone(ControlNodeIotlab):
```
Comment (rienafairefr) [inline gateway_code/control_nodes/cn_iotlabstalone/__init__.py]: stalone everywhere, maybe it's not a typo in the end ?

File: gateway_code/control_nodes/cn_iotlab/__init__.py (hunk 2 above)
Comment (rienafairefr) [reply gateway_code/control_nodes/cn_iotlab/__init__.py]: Well, it's actually breaking the linter, because the status method should be static (we don't use self in it). either ignore it (add appropriate pylint: ignore),

File: gateway_code/control_nodes/cn_iotlabstalone/__init__.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    TYPE = 'iotlabstalone'
+    FEATURES = ['leds',
+                'radio']
+    OPENOCD_CFG_FILE = static_path('iot-lab-cn-stalone.cfg')
+
+    def __init__(self, node_id, default_profile):
```
Comment (rienafairefr) [inline gateway_code/control_nodes/cn_iotlabstalone/__init__.py]: not needed, because your class is a subclass of ControlNodeIotlab

File: gateway_code/control_nodes/cn_iotlabstalone/__init__.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        super(ControlNodeIotlabstalone, self).__init__(node_id, default_profile)
+
+    @logger_call("Control node : profile configuration")
+    def configure_profile(self, profile=None):
+        """ Configure the given profile on the control node """
+        LOGGER.info('Configure profile on Control Node')
+        self.profile = profile or self.default_profile
+        ret_val = 0
+
```
Comment (rienafairefr) [inline gateway_code/control_nodes/cn_iotlabstalone/__init__.py]: plenty of weird lines with spaces at the end

File: tests_utils/test-cn-stalone-Docker.sh [hunk 6]
```diff
@@ -0,0 +1,51 @@
+#! /bin/bash -x
```
Comment (rienafairefr) [inline tests_utils/test-cn-stalone-Docker.sh]: this should be rewritten to take into account the Makefile that has been added in the meantime

**Review by rienafairefr — CHANGES_REQUESTED** [review]
Still some work to do :-) :+1:

**Comment by rdagher** [comment]
Gr8. I made some updates, and it works on my setup. Check it out.

File: gateway_code/control_nodes/cn_iotlabstalone/__init__.py (hunk 1 above)
Comment (aabadie) [reply gateway_code/control_nodes/cn_iotlabstalone/__init__.py]: and why not simply `cn_iotlabm3` ?

File: gateway_code/utils/ftdi_check.py [hunk 7]
```diff
@@ -21,17 +21,73 @@
...
+        found = (description == ftdi_parse_device_description(lines[i + 2]))
+        if found:
+            break
+    return found
+
+
+if __name__ == "__main__":
+    print ftdi_check('control', '2232')
+    print ftdi_check('control', '2232', 'ControlNode')
+    import doctest
```
Comment (aabadie) [inline gateway_code/utils/ftdi_check.py]: I think this is already called by nose (and soon pytest)

File: gateway_code/control_nodes/cn_iotlabstalone/__init__.py (hunk 1 above)
Comment (rdagher) [reply gateway_code/control_nodes/cn_iotlabstalone/__init__.py]: OK, Changed the name to `cn_iotlabm3 `.

File: gateway_code/utils/ftdi_check.py (hunk 7 above)
Comment (rdagher) [reply gateway_code/utils/ftdi_check.py]: Ok. Removed the main.

**Review by rienafairefr — APPROVED** [review]

File: Makefile [hunk 8]
```diff
@@ -46,9 +46,11 @@ run: setup-cfg-dir
 	docker run -it --rm \
 		-v $(PWD):/shared \
 		-v /dev/iotlab:/dev/iotlab \
+		-v /dev/ttyCN:/dev/ttyCN \
 		-v /tmp/cfg_dir:/var/local/config \
 		-p $(PORT):8080 \
 		-p $(SERIAL_PORT):20000 \
+		--name "gateway_test" \
```
Comment (rienafairefr) [inline Makefile]: Yeah I think we shouldn't force docker container names, esp. if you re-do a docker run with the same name it will error. I removed 'gateway-test' name in 8c7f540475d736384891f793e30256533a0fd2b7 when deleting the docker-run

File: Makefile [hunk 9]
```diff
@@ -46,9 +46,11 @@ run: setup-cfg-dir
 	docker run -it --rm \
 		-v $(PWD):/shared \
 		-v /dev/iotlab:/dev/iotlab \
+		-v /dev/ttyCN:/dev/ttyCN \
```
Comment (rienafairefr) [inline Makefile]: @aabadie  mentioned it, but what happens when the host doesn't have /dev/ttyCN ?

File: tests_utils/test-cn-m3-Docker.sh [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+# - Docker container named 'gateway_test' attached to a standalone M3 OpenNode and standalone M3 ControlNode
+# - The ControlNode board's FTDI eeprom has been set with the name ControlNode (using the IotLab ftdi-eeprom-config tool)
+
+# Experiment profile (sniffer) and firmware
+profile="../gateway_code/tests/profiles/radio_sniffer_3.json"
+firmware="../gateway_code/static/m3_autotest.elf"
+
+# Mocked user workspace and experiment folders (usually mounted by nfs)
+# here, they are created in the (alreadyà running container
+user="test"
```
Comment (rienafairefr) [inline tests_utils/test-cn-m3-Docker.sh]: you should use or modify one of the makefile targets (I think `make run` will do what you need, just add a -v mapping a folder in /tmp to /iotlab/users/.../experiment)

**Review by rienafairefr — CHANGES_REQUESTED** [review]
the docker script is still messy with the docker exec, try the make run I just modified

**Comment by rdagher** [comment]
I don't understand why you want to avoid explicit container name, but I do understand the use of volumes to retrieve the experiment output. 
Updated the [test-cn-m3-Docker.sh](https://github.com/rdagher/iot-lab-gateway/blob/stalone_cn_m3/tests_utils/test-cn-m3-Docker.sh) to first run the server as suggested and dump sniffing result from the shared volume `/tmp/exp_dir/sniffer/`.

File: Makefile (hunk 9 above)
Comment (rdagher) [reply Makefile]: What happens is that the Docker daemon creates a folder `/dev/ttyCN` and mounts it !
Of course this is not clean. What about moving the symbolic link `ttyCN` to the `/dev/iotlab`folder? 
I can give it a shot (update rules and code) and test) if you agree...

**Comment by rienafairefr** [comment]
The problem with named containers is that once a name is taken by a container you can't create another one with the same name, even if the conflict comes from a stopped container, so re-launching stuff can fail

File: Makefile (hunk 9 above)
Comment (rienafairefr) [reply Makefile]: Yeah I think the best would be to put ttyCN in /dev/iotlab. 
in udev SYMLINK+="ttyCN[separator]iotlab/ttyCN" would create both symbolic links, I think separator is a comma not sure

Comment (aabadie) [reply Makefile]: I would prefer adding some Makefile logic that sets a variable with mapping if requested by the user:
```
make USE_CN=1 integration-test
```

In the makefile, test that USE_CN is equal to 1, default value is 0:
```mk
USE_CN ?=0

DOCKER_CN_MAPPING=
ifeq (1,$(USE_CN))
  DOCKER_CN_MAPPING = -v /dev/ttyCN:/dev/ttyCN
endif
```

And then in the integration-test target:
```mk
docker run -it --rm \
 		-v $(PWD):/shared \
 		-v /dev/iotlab:/dev/iotlab \
		$(DOCKER_CN_MAPPING) \
```

Not tested, but you have the idea.

Comment (rdagher) [reply Makefile]: @aabadie I see the idea, the logic is actually in the variable `CONTROL_NODE_TYPE`. I am pushing a fix.

**Comment by rienafairefr** [comment]
with latest additions, looks good to me

**Review by rienafairefr — APPROVED** [review]

**Comment by rienafairefr** [comment]
@aabadie, I think I'm merging this today unless you have last-minute objections

**Comment by aabadie** [comment]
FYI, this PR broke the integration tests. After digging into the issue on our CI, it appears that the control node on legacy gateways was not working as it should. In the end, the problem was that this PR modifies the firmware that is running on the control node and this breaks the control node on legacy IoT-LAB gateways.

So I pushed a commit that revert the original control node firmware. If you want to use a specific one for standalone M3 CN, please provide it separately. Ideally, you should also provide the source somewhere (not in this repo). Thanks !

**Comment by rdagher** [comment]
@aabadie I had to recompile the firmware of the control node with the latest version from openlab. The reason I had to do that is that I had some issues with the control node interface, somme commands were missing I think, but i will test this again with the firmware you reverted. 
BTW, in production, this same firmware is used to flash the CN or is there any repo for CN images?

**Comment by aabadie** [comment]
I'm not 100% sure but I think the source code of the application that runs on the control node is this [here](https://github.com/iot-lab/openlab/tree/master/appli/iotlab/control_node/control_node).
I don't think it's possible to use exact same application for the legacy CN and an M3 base CN, because the M3 base CN won't be able to control the alim of the open node.

Maybe you forgot to pass the `IOTLAB_CN` build flag when you rebuilt the new firmware version.
Now I think the best is to provide the firmware that run the M3 as CN in a separate file and adapt the `ControlNodeIotlabm3` python class accordingly.
